cpp/src/reductions/reduction_dispatcher.cuh

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CUDF_REDUCTION_DISPATCHER_CUH
+#define CUDF_REDUCTION_DISPATCHER_CUH
+
+#include "reduction_functions.cuh"
+
+namespace cudf {
+namespace reductions {
+
+// make an column iterator
+template<bool has_nulls, typename T_in, typename T_out, typename Op>
+auto make_iterator(const gdf_column &input, T_out identity, Op op)
+{
+    return cudf::make_iterator<has_nulls, T_in, T_out>(input, identity);
+}
+
+// make an column iterator, specialized for `cudf::reductions::ReductionSumOfSquares`
+template<bool has_nulls, typename T_in, typename T_out>
+auto make_iterator(const gdf_column &input, T_out identity,
+    cudf::reductions::ReductionSumOfSquares op)
+{
+    auto it_raw = cudf::make_iterator<has_nulls, T_in, T_out>(input, identity);
+    return thrust::make_transform_iterator(it_raw,
+        cudf::transformer_squared<T_out>{});
+}
+
+// Reduction for 'sum', 'product', 'min', 'max', 'sum of squares'
+// which directly compute the reduction by a single step reduction call
+template<typename T_in, typename T_out, typename Op, bool has_nulls>
+void ReduceOp(const gdf_column *input,
```

Comment:
Contrast this to the "multi-step" or "compound" reduction with something like:
```suggestion
void simple_reduction(const gdf_column *input,
```

Reply:
Also note we are trying to standardize on `type_name` naming instead of `TypeName` naming.